File: JavaEE/Pharmacy-EJB/ejbModule/de/java/ejb/jms/OrderMarshaller.java
package de.java.ejb.jms;

import static de.java.util.DateFormatter.format;
import static de.java.util.MessageDelimiter.DELIMITER;
import static de.java.util.MessageDelimiter.ORDER_DELIMITER;

import java.util.Collection;
import java.util.Iterator;

import de.java.domain.OrderState;
import de.java.domain.Position;
import de.java.domain.ReplenishmentOrder;

public class OrderMarshaller {

  public static String marshalAll(Collection<ReplenishmentOrder> orders) {
    String marshalledResult = "";
    Iterator<ReplenishmentOrder> iterator = orders.iterator();
    while (iterator.hasNext()) {
      ReplenishmentOrder order = iterator.next();
      marshalledResult += order.getId();
      marshalledResult += DELIMITER;
      marshalledResult += order.getState();
      if (order.getState() == OrderState.ORDERED
          || order.getState() == OrderState.FINISHED) {
        marshalledResult += DELIMITER;
        marshalledResult += format(order.getExpectedDelivery());
      }
      if (order.getState() == OrderState.FINISHED) {
        marshalledResult += DELIMITER;
        marshalledResult += format(order.getActualDelivery());
      }
      if (iterator.hasNext()) {
        marshalledResult += ORDER_DELIMITER;
      }
    }
    return marshalledResult;
  }

  public static String marshalSingle(ReplenishmentOrder order) {
    String marshalledResult = "";
    marshalledResult += order.getId();
    marshalledResult += DELIMITER;
    marshalledResult += order.getState();
    marshalledResult += marshalPositions(order.getPositions());
    if (order.getState() == OrderState.ORDERED
        || order.getState() == OrderState.FINISHED) {
      marshalledResult += DELIMITER;
      marshalledResult += format(order.getExpectedDelivery());
    }
    if (order.getState() == OrderState.FINISHED) {
      marshalledResult += DELIMITER;
      marshalledResult += format(order.getActualDelivery());
    }
    return marshalledResult;
  }

  private static String marshalPositions(Collection<Position> positions) {
    String marshalledResult = "";
    marshalledResult += DELIMITER;
    marshalledResult += positions.size();
    for (Position p : positions) {
      marshalledResult += DELIMITER;
      marshalledResult += p.getReplenishedDrug().getPzn();
      marshalledResult += DELIMITER;
      marshalledResult += p.getQuantity();
    }
    return marshalledResult;
  }

}

File: DotNET/JmsClient/OrderMarshaller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Pharmacy.BusinessLayer.Data;

namespace JmsClient
{
    class OrderMarshaller
    {
        public static char DELIMITER = ';';
        public static string DATE_FORMAT = "yyyy-MM-dd HH:mm";

        public static String MarshalAll(ICollection<ReplenishmentOrder> orders) {
            return String.Join("\n", orders.Select(item => MarshalSingleForAll(item)));
        }

        private static String MarshalSingleForAll(ReplenishmentOrder order)
        {
            String marshalledResult = "";
            marshalledResult += order.Id;
            marshalledResult += DELIMITER;
            marshalledResult += order.State.ToString().ToUpper();
            if (order.State == OrderState.Ordered
                || order.State == OrderState.Finished)
            {
                marshalledResult += DELIMITER;
                marshalledResult += Format(order.ExpectedDelivery);
            }
            if (order.State == OrderState.Finished)
            {
                marshalledResult += DELIMITER;
                marshalledResult += Format(order.ActualDelivery);
            }
            return marshalledResult;
        }

        private static String Format(DateTime? dateTime)
        {
            if (dateTime == null)
            {
                throw new ArgumentNullException("no datetime passed in to format");
            }
            return ((DateTime)dateTime).ToString(DATE_FORMAT);
        }


        public static String MarshalSingle(ReplenishmentOrder order)
        {
            String marshalledResult = "";
            marshalledResult += order.Id;
            marshalledResult += DELIMITER;
            marshalledResult += order.State.ToString().ToUpper();
            marshalledResult += MarshalPositions(order.Positions);
            if (order.State == OrderState.Ordered
               || order.State == OrderState.Finished)
            {
                marshalledResult += DELIMITER;
                marshalledResult += Format(order.ExpectedDelivery);
            }
            if (order.State == OrderState.Finished)
            {
                marshalledResult += DELIMITER;
                marshalledResult += Format(order.ActualDelivery);
            }
            return marshalledResult;
        }

        private static String MarshalPositions(ICollection<Position> positions)
        {
            String marshalledResult = "";
            marshalledResult += DELIMITER;
            marshalledResult += positions.Count;
            foreach (Position p in positions) {
              marshalledResult += DELIMITER;
              marshalledResult += p.DrugPZN;
              marshalledResult += DELIMITER;
              marshalledResult += p.Quantity;
            }
            return marshalledResult;
        }
    }
}
